h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -3579,177 +3790,189 @@ private void doCompute() {
-                          "] = " + _parms._alpha[alphaInd] + ", improvement on test = " + s));
+
+        if (_parms._HGLM) { // add w, augZ, etaOld and random columns to response for easy access inside _dinfo
+          addWdataZiEtaOld2Response();
+        }
+
+        double oldDevTrain = nullDevTrain;
+        double oldDevTest = nullDevValid;
+        double[] devHistoryTrain = new double[5];
+        double[] devHistoryTest = new double[5];
+
+        if (!_parms._HGLM) {  // only need these for non HGLM
+          _ginfoStart = GLMUtils.copyGInfo(_state.ginfo());
+          _betaDiffStart = _state.getBetaDiff();
+        }
+
+        if (_parms.hasCheckpoint()) { // restore _state parameters
+          _state.copyCheckModel2State(_model, _gamColIndices);
+          if (_model._output._submodels.length == 1)
+            _model._output._submodels = null; // null out submodel only for single alpha/lambda values
+        }
+
+        if (!_parms._lambda_search & !_parms._HGLM)
+          updateProgress(false);
+
+        // alpha, lambda search loop
+        int alphaStart = 0;
+        int lambdaStart = 0;
+        int submodelCount = 0;
+        if (_parms.hasCheckpoint() && _model._output._submodels != null) {  // multiple alpha/lambdas or lambda search
+          submodelCount = Family.gaussian.equals(_parms._family) ? _model._output._submodels.length
+                  : _model._output._submodels.length - 1;
+          alphaStart = submodelCount / _parms._lambda.length;
+          lambdaStart = submodelCount % _parms._lambda.length;
+        }
+        _model._output._lambda_array_size = _parms._lambda.length;
+        for (int alphaInd = alphaStart; alphaInd < _parms._alpha.length; alphaInd++) {
+          _state.setAlpha(_parms._alpha[alphaInd]);   // loop through the alphas
+          if ((!_parms._HGLM) && (alphaInd > 0) && !_checkPointFirstIter) // no need for cold start during the first iteration
+            coldStart(devHistoryTrain, devHistoryTest);  // reset beta, lambda, currGram
+          for (int i = lambdaStart; i < _parms._lambda.length; ++i) {  // for lambda search, can quit before it is done
+            if (_job.stop_requested() || (timeout() && _model._output._submodels.length > 0))
+              break;  //need at least one submodel on timeout to avoid issues.
+            if (_parms._max_iterations != -1 && _state._iter >= _parms._max_iterations)
+              break;  // iterations accumulate across all lambda/alpha values when coldstart = false
+            if ((!_parms._HGLM && (_parms._cold_start || (!_parms._lambda_search && _parms._cold_start))) && (i > 0)
+                    && !_checkPointFirstIter) // default: cold_start for non lambda_search
+              coldStart(devHistoryTrain, devHistoryTest);
+            Submodel sm = computeSubmodel(submodelCount, _parms._lambda[i], nullDevTrain, nullDevValid);
+            if (_checkPointFirstIter)
```

Comment:
`if (_checkPointFirstIter)` seems unnecessary. You can keep just the next line(`_checkPointFirstIter = false;`).

Reply:
Logically you are correct, but I found this saying that checking it is much faster than blunt writing:
![image](https://github.com/h2oai/h2o-3/assets/7231712/8f15a863-94fd-4a99-a5e7-93ca6d1fa2e0)

https://stackoverflow.com/questions/23228359/is-checking-a-boolean-faster-than-setting-a-boolean-in-java#:~:text=Result%20(under%20these%20conditions%2C%20and,it%20often%20if%20not%20necessary.&text=With%20checking%2C%20it%20takes%20about,otherwise%20happen%20unnecessarily%20very%20often.

Reply:
However, I also found contradicting views...

Reply:
Thank you for taking time to look into performance implications but I don't think there would be any measurable effect - it is run `_lambda.length * _alpha.length` times. What is the typical size of `_alpha[]` and `_lambda[]` ?

Never mind that, it's not really an issue, so feel free to ignore this.